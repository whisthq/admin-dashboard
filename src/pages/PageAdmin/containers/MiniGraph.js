import React, { Component } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCircleNotch } from '@fortawesome/free-solid-svg-icons'
import { LineChart, Line, ResponsiveContainer, YAxis, Tooltip } from 'recharts'

import 'react-tabs/style/react-tabs.css'

import 'static/App.css'

class MiniGraph extends Component {
    render() {
        if (
            !this.props.filename ||
            (this.props.log_analysis &&
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ] &&
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender] &&
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender][this.props.metric])
        ) {
            if (
                !this.props.filename ||
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender][this.props.metric].output.length < 2
            ) {
                return (
                    <div
                        style={{
                            paddingLeft: 10,
                            width: 300,
                        }}
                    >
                        <div
                            style={{
                                fontWeight: 'bold',
                                marginBottom: 10,
                                marginTop: 10,
                            }}
                        >
                            {this.props.title}
                        </div>
                        <div style={{ fontSize: 13, marginTop: 10 }}>
                            Not enough data to plot
                        </div>
                    </div>
                )
            } else {
                return (
                    <div>
                        <div
                            style={{
                                fontWeight: 'bold',
                                marginBottom: 10,
                                marginTop: 10,
                                paddingLeft: 10,
                            }}
                        >
                            {this.props.title}
                        </div>
                        <div
                            style={{
                                position: 'relative',
                                right: 30,
                                marginTop: 10,
                            }}
                        >
                            <ResponsiveContainer width={'100%'} height={150}>
                                <LineChart
                                    data={
                                        this.props.log_analysis[
                                            this.props.username.concat(
                                                '_',
                                                this.props.connection_id
                                            )
                                        ][this.props.sender][this.props.metric]
                                            .output
                                    }
                                    syncId="usageChart"
                                >
                                    <Line
                                        type="monotone"
                                        dataKey="value"
                                        dot={false}
                                        strokeWidth={2}
                                        stroke="#8884d8"
                                    />
                                    <YAxis
                                        dataKey="value"
                                        axisLine={false}
                                        tickLine={false}
                                        allowDecimals={false}
                                        tick={{
                                            fontSize: 12,
                                        }}
                                        tickFormatter={(element) => {
                                            if (Number(element) > 10000) {
                                                return (
                                                    (
                                                        Number(element) / 1000
                                                    ).toString() + 'K'
                                                )
                                            } else {
                                                return element.toString()
                                            }
                                        }}
                                    />
                                    <Tooltip
                                        contentStyle={{
                                            border: 'none',
                                            fontSize: 14,
                                        }}
                                        labelFormatter={(element) => {
                                            return 'Time: '.concat(
                                                this.props.log_analysis[
                                                    this.props.username.concat(
                                                        '_',
                                                        this.props.connection_id
                                                    )
                                                ][this.props.sender][
                                                    this.props.metric
                                                ].output[element].time
                                            )
                                        }}
                                        formatter={(value) =>
                                            value
                                                .toFixed(2)
                                                .toString()
                                                .concat(' ', this.props.unit)
                                        }
                                    />
                                </LineChart>
                            </ResponsiveContainer>
                        </div>
                    </div>
                )
            }
        } else {
            return (
                <div
                    style={{
                        paddingLeft: 10,
                        width: 300,
                    }}
                >
                    <div
                        style={{
                            fontWeight: 'bold',
                            marginBottom: 10,
                            marginTop: 10,
                        }}
                    >
                        {this.props.title}
                    </div>
                    <div style={{ fontSize: 13 }}>
                        <FontAwesomeIcon
                            icon={faCircleNotch}
                            spin
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                                color: '#4b3ba8',
                            }}
                        />
                        Calculating {this.props.title}
                    </div>
                </div>
            )
        }
    }
}

export default MiniGraph
