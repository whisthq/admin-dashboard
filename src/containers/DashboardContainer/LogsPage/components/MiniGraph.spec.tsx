import React from 'react'
import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { LineChart } from 'recharts'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

import { MiniGraph } from './MiniGraph'

Enzyme.configure({ adapter: new Adapter() })

/*
The MiniGraph is a small graph that appears inside each log element in the big infinte scroll. It takes in
its plotted data from props (which if you go up far enough comes from Redux + middleware).

It needs at least two datapoints to plot since it is a line graph. If it can plot in has a line graph chart.
Otherwise, if there is no data it shows the icon and a loading message.
 */
describe('<MiniGraph />', () => {
    it('displays an icon and a message if it gets empty/null props', () => {
        const wrapper = shallow(
            <MiniGraph
                // buncha dummy data
                filename={false} // should short-circuit on the first if statement
                dispatch={(items: any) => {
                    // mock
                }}
                title="title"
                username="user"
                sender="sender"
                connection_id="id"
                metric="metric"
                logAnalysis={null}
            />
        )

        expect(wrapper.find(LineChart)).toHaveLength(0)
        expect(wrapper.find(FontAwesomeIcon)).toHaveLength(1)
    })

    it('displays a message instead of a plot if there is not enough data', () => {
        const wrapper = shallow(
            <MiniGraph
                // buncha dummy data
                dispatch={(items: any) => {
                    // mock
                }}
                title="title"
                filename="filename"
                username="user"
                sender="sender"
                connection_id="id"
                metric="metric"
                logAnalysis={{
                    user_id: {
                        sender: {
                            metric: {
                                output: [1], // not enough data to plot
                            },
                        },
                    },
                }}
            />
        )

        expect(wrapper.find(LineChart)).toHaveLength(0)
        expect(wrapper.find('div')).toHaveLength(3) // a div containing one with the title and one with a message
    })

    it('displays a plot if it gets chartable props', () => {
        const wrapper = shallow(
            <MiniGraph
                // buncha dummy data
                filename="filename"
                dispatch={(items: any) => {
                    // mock
                }}
                title="title"
                username="user"
                sender="sender"
                connection_id="id"
                metric="metric"
                logAnalysis={{
                    user_id: {
                        sender: {
                            metric: {
                                output: [1, 2, 3],
                            },
                        },
                    },
                }}
            />
        )

        expect(wrapper.find(LineChart)).toHaveLength(1)
    })
})
