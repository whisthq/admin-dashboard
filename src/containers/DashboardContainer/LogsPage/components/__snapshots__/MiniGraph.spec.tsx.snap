// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MiniGraph /> renders correctly 1`] = `
<div
  style={
    Object {
      "paddingLeft": 10,
      "width": 300,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": "bold",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    title
  </div>
  <div
    style={
      Object {
        "fontSize": 13,
        "marginTop": 10,
      }
    }
  >
    Not enough data to plot
  </div>
</div>
`;
